internal/httptransport/httptransport.go

```diff
@@ -0,0 +1,250 @@
+}
+
+// RoundTrip implements Transport.RoundTrip.
+func (txp HeaderAdder) RoundTrip(req *http.Request) (*http.Response, error) {
+	if req.Header.Get("User-Agent") == "" {
+		if txp.UserAgent != "" {
+			req.Header.Set("User-Agent", txp.UserAgent)
+		} else {
+			req.Header["User-Agent"] = nil // disable sending user agent
+		}
+	}
+	host := req.Host
+	if host == "" {
+		host = req.URL.Host
+	}
+	req.Header.Set("Host", host) // have it set explicitly
+	return txp.Transport.RoundTrip(req)
+}
+
+// SnapshotSaver saves a snapshot of the response body
+type SnapshotSaver struct {
+	Transport
+	SnapshotSize int64
+	snapshots    []BodySnapshot
+	mu           sync.Mutex
+}
+
+// Snapshots returns the saved body snapshots
+func (txp *SnapshotSaver) Snapshots() []BodySnapshot {
+	txp.mu.Lock()
+	snapshots := txp.snapshots
+	txp.snapshots = nil
+	txp.mu.Unlock()
+	return snapshots
+}
+
+// RoundTrip implements Transport.RoundTrip.
+func (txp *SnapshotSaver) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp, err := txp.Transport.RoundTrip(req)
+	if err != nil {
+		return nil, err
+	}
+	size := txp.SnapshotSize
+	if size <= 0 {
+		size = 1 << 18
+	}
+	reader := io.LimitReader(resp.Body, size)
+	data, err := ioutil.ReadAll(reader)
+	snapshot := BodySnapshot{Data: string(data), Err: err, ReadCloser: resp.Body,
+		Time: time.Now(), URL: req.URL.String()}
+	txp.mu.Lock()
+	txp.snapshots = append(txp.snapshots, snapshot)
+	txp.mu.Unlock()
+	resp.Body = snapshot
+	return resp, nil
+}
+
+// BodySnapshot is a snapshot of the response body
+type BodySnapshot struct {
+	io.ReadCloser
+	Data string
+	Err  error
+	Time time.Time
+	URL  string
+}
+
+// Read implements ReadCloser.Reader.Read
+func (body BodySnapshot) Read(p []byte) (int, error) {
+	if body.Err != nil {
+		return 0, body.Err
+	}
+	return body.ReadCloser.Read(p)
+}
+
+// EventsSaver saves events occurring during the round trip
+type EventsSaver struct {
+	Transport
+	events []Events
+	mu     sync.Mutex
+}
+
+// RoundTrip implements Transport.RoundTrip.
+func (txp *EventsSaver) RoundTrip(req *http.Request) (*http.Response, error) {
+	var (
+		rte Events
+		mu  sync.Mutex
+	)
+	tracer := &httptrace.ClientTrace{
+		GetConn: func(hostPort string) {
+			mu.Lock()
+			rte.GetConnTime = time.Now()
+			rte.GetConnAddress = hostPort
+			mu.Unlock()
+		},
+		GotConn: func(info httptrace.GotConnInfo) {
+			mu.Lock()
+			rte.GotConnTime = time.Now()
+			rte.GotConnAddress = info.Conn.LocalAddr().String()
+			mu.Unlock()
+		},
+		GotFirstResponseByte: func() {
+			mu.Lock()
+			rte.GotFirstResponseByteTime = time.Now()
+			mu.Unlock()
+		},
+		WroteHeaders: func() {
+			mu.Lock()
+			rte.WroteHeadersTime = time.Now()
+			mu.Unlock()
+		},
+		WroteRequest: func(httptrace.WroteRequestInfo) {
+			mu.Lock()
+			rte.WroteRequestTime = time.Now()
+			mu.Unlock()
+		},
+	}
+	rte.URL = req.URL.String()
+	rte.RoundTripStartTime = time.Now()
+	req = req.WithContext(httptrace.WithClientTrace(req.Context(), tracer))
```

Comment:
This creates a new context. While the callbacks are merged, this context only applies to this request. This is nice because in principle a user could choose to set their own callbacks (which in practice won't happen I think).